### PR TITLE
Sort user comments on profile page

### PR DESCRIPTION
**Pull request by yoannLafore**
Hello,
This PR aims to sort the user comments on the profile page (issue #300).

Concretelly, this PR does the following:

**Add `publicationTime` field to `UserCommentData`:**
A new field `publicationTime` has been added to the `UserCommentData` to allow for the sorting.
Thus, the repository `CommentRepositoryService` as well as the tests and mocks for `UserCommentData` have been updated to take into account this new field.

**User comments view model:**
The view model to display the user comments will now sort the comments from latest to oldest before exposing them.
Thus, the latest comments of the user wil appear at the top of the list on the profile page.

**Tests:**
This PR provides a test for the ordering of the user comments exposed by the view model.

#### Live testing
You can test this feature by doing the following:
In the application, add a new comment under any post then go into your profile page in the comment tab.
You should then see the newly added comment at the top. You can add more comments and you should see them appearing sorted from latest (top) to oldest (bottom) in the profile page.

#### Acceptance criteria:
- [x] Sort the user's comments within the user profile, displaying the latest comments first.

**Note:**
Once this PR is merged, the comments on the database will need to be cleared because the structure `UserCommentData` has changed.

Thank you for reviewing this PR and I look forward to your feedback :+1:

### CODE REVIEW

**Review by camillelnne — COMMENTED** [review]
Very good job!

I get this bug when testing it on my phone (I think I am supposed to have two comments)

image

It might be because you added the publication time to comments but the change is not retroactive, so older comments don't have it. I don't have this problem with newly created comments.
You mentioned it in the description, but just to be sure, can you confirm this is the cause of the bug?

File: lib/viewmodels/user_comments_view_model.dart [hunk 1]
```diff
@@ -17,7 +17,15 @@ class UserCommentViewModel extends AutoDisposeAsyncNotifier<UserCommentsState> {
     final user = ref.watch(validLoggedInUserIdProvider);
 
     final commentsFirestore = await userCommentRepository.getUserComments(user);
-    final comments = commentsFirestore.map((comment) {
+
+    // Sort the comments from latest to oldest
+    final sortedCommentsFirestore = commentsFirestore.toList()
+      ..sort(
```
[the remark this answers is not in the document]
Comment (yoannLafore) [reply lib/viewmodels/user_comments_view_model.dart]: You are totally right !
There are no particular reason for this choice.

And I actually think that it is better to use the same function for both for consistency.
Thus I make use of `sorted` in 8dabb917fa59363dbc08f98fc4c1785db7bf695d.

File: test/viewmodels/user_comment_view_model_test.dart [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        overrides: [
+          firestoreProvider.overrideWithValue(firestore),
+          loggedInUserIdProvider.overrideWithValue(userId),
+        ],
+      );
+
+      commentRepository = container.read(commentRepositoryServiceProvider);
+    });
+
+    group("Ordering", () {
```
[the remark this answers is not in the document]
Comment (yoannLafore) [reply test/viewmodels/user_comment_view_model_test.dart:40]: That's a good idea to test such an edge case !

Done in b604cbb8f8cce25b0ddb46087934294b49a1eac5.

**Comment by yoannLafore** [comment]
Thank you for your great suggestions :tada: 

Regarding the bug that you are encountering, you are right, this is because old `UserComment`s don't have the `publicationTime` field in the database.
To solve this issue, the simplest way is to clean the database from the old user comments.
I'll do it just before merging the PR.

File: test/viewmodels/user_comment_view_model_test.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (camillelnne) [reply test/viewmodels/user_comment_view_model_test.dart:40]: Maybe you can also do for a single comment like you did for the post :) (or maybe I'm obsessed with symmetry haha)

[the remark this answers is not in the document]
Comment (yoannLafore) [reply test/viewmodels/user_comment_view_model_test.dart:40]: That's a good idea indeed :)

Added in 5d3af3df14a684fc9deda4238b2e4f4ccdd0a05a.

[the remark this answers is not in the document]
Comment (camillelnne) [reply test/viewmodels/user_comment_view_model_test.dart:40]: Thanks!

**Review by camillelnne — APPROVED** [review]
Thanks for the changes!
LGTM

**Comment by yoannLafore** [comment]
After discussion with the team, we decided to add a simple backward compatibility mechanism with the old database structure.
Thus, if the `publicationTime` is not present, it will be considered to be `Timestamp(0)` and those comments will be put at the bottom of the list.

This is added in 4f5388012d31d36072046649080a6d88a86ff441.

**Review by camillelnne — APPROVED** [review]
Good job on the backward compatibility, will definitely save you some trouble and is a good practice :)
Everything seems to work perfectly.
